File: L3/Android/Java/RecyclerView-java/First-java/app/src/main/java/com/example/jamian/first/MyMessageRecyclerViewAdapter.java
package com.example.jamian.first;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.example.jamian.first.MessagesFragment.OnListFragmentInteractionListener;

import androidx.recyclerview.widget.RecyclerView;

import java.util.Date;

/**
 * {@link RecyclerView.Adapter} that can display a {@link Message} and makes a call to the
 * specified {@link OnListFragmentInteractionListener}.
 */
public class MyMessageRecyclerViewAdapter extends RecyclerView.Adapter<MyMessageRecyclerViewAdapter.ViewHolder> {

    private final ListeMessages mValues;
    private final OnListFragmentInteractionListener mListener;

    public MyMessageRecyclerViewAdapter(ListeMessages items, OnListFragmentInteractionListener listener) {
        mValues = items;
        mListener = listener;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.fragment_message, parent, false);
        return new ViewHolder(view);
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, int position) {
        holder.mItem = mValues.get(position);
        holder.mIdView.setText(mValues.get(position).getAuthor());
        String textContent = mValues.get(position).getContenu();
        if (textContent.length() > 50) textContent = textContent.substring(0,47)+"...";
        holder.mContentView.setText(textContent);

        holder.mView.setOnClickListener(v -> {
            if (null != mListener) {
                // Notify the active callbacks interface (the activity, if the
                // fragment is attached to one) that an item has been selected.
                mListener.onListFragmentInteraction(holder.mItem);
            }
        });
    }

    @Override
    public int getItemCount() {
        return mValues.size();
    }

    public void ajouteMessage(int id, Date date, String author, String contenu)
    {
        mValues.ajouteMessage(id,date,author,contenu);
        this.notifyItemInserted(mValues.size()-1);
    }
    public void supprimeMessage(int id)
    {
        int msgPos = mValues.deleteMessageFromId(id);
        if (msgPos >=0)
        {
            this.notifyItemRemoved(msgPos);
            this.notifyItemRangeChanged(msgPos,mValues.size());
        }
    }

    public class ViewHolder extends RecyclerView.ViewHolder {
        public final View mView;
        public final TextView mIdView;
        public final TextView mContentView;
        public Message mItem;

        public ViewHolder(View view) {
            super(view);
            mView = view;
            mIdView = (TextView) view.findViewById(R.id.id);
            mContentView = (TextView) view.findViewById(R.id.content);
        }

        @Override
        public String toString() {
            return super.toString() + " '" + mContentView.getText() + "'";
        }
    }
}

File: L3/Android/RecyclerView-kt/First/app/src/main/java/com/example/jamian/first/MyMessageRecyclerViewAdapter.kt
package com.example.jamian.first

import com.example.jamian.first.MessagesFragment.OnListFragmentInteractionListener
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import android.widget.TextView
import java.util.Date

/**
 * [RecyclerView.Adapter] that can display a [Message] and makes a call to the
 * specified [OnListFragmentInteractionListener].
 */
class MyMessageRecyclerViewAdapter(private val mValues: ListeMessages, private val mListener: OnListFragmentInteractionListener?) : RecyclerView.Adapter<MyMessageRecyclerViewAdapter.ViewHolder>() {
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.fragment_note, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.mItem = mValues[position]
        holder.mAuthorView.text = mValues[position]!!.author
        var textContent = mValues[position]!!.msg
        if (textContent.length > 50) textContent = textContent.substring(0, 47) + "..."
        holder.mMsgView.text = textContent
        holder.mView.setOnClickListener { mListener?.onListFragmentInteraction(holder.mItem) }
    }

    override fun getItemCount(): Int {
        return mValues.size()
    }

    fun ajouteMessage(id:Int,date:Date,author: String, contenu: String) {
        mValues.ajouteMessage(id, date, contenu)
        notifyItemInserted(mValues.size() - 1)
    }

    fun supprimeMessage(id: Int) {
        val notePos = mValues.deleteMessageFromId(id)
        if (notePos >= 0) {
            notifyItemRemoved(notePos)
            this.notifyItemRangeChanged(notePos, mValues.size())
        }
    }

    inner class ViewHolder(val mView: View) : RecyclerView.ViewHolder(mView) {
        val mAuthorView: TextView
        val mMsgView: TextView
        val mDateView: TextView
        var mItem: Message? = null
        override fun toString(): String {
            return super.toString() + " '" + mMsgView.text + "'"
        }

        init {
            mAuthorView = mView.findViewById<View>(R.id.id) as TextView
            mMsgView = mView.findViewById<View>(R.id.content) as TextView
            mDateView = mView.findViewById<View>(R.id.date) as TextView
        }
    }
}